Simplify nested Min and Max patterns.

navahgar opened the pull request:
Improve simplification of nested Min and Max patterns. 

Specifically, handles the following pattern simplications:
  * `Max(A, Max(A, Const)) => Max(A, Const)`
  * `Max(Min(A, B), Min(A, C)) => Min(A, Max(B, C))`
  * `Max(Const, Max(A, OtherConst) => Max(A, Max(Const, OtherConst))`
     - This case can have an arbitrarily long chain of Max ops. For example: `Max(5, Max(x, Max(y, Max(z, 8)))) => Max(Max(Max(x, 8), y), z)`

Similarly, for the case of Min as well.

navahgar committed:
Simplify nested Min and Max patterns.

nickgg reviewed (changes requested):
Ok so functionally this looks great, and I like the test coverage (although I think you can commonalize a lot of the code), but I think we should refactor this before landing - there's too much copied code that can be eliminated. It's worth aiming for as clear and simple an implementation as possible here because the simplifier is a black box to most users and we want to smooth out any maintenance as much as we can.

navahgar commented on test/cpp/tensorexpr/test_simplify.cpp:
```diff
@@ -1401,6 +1401,738 @@ void testSimplifySymbolicMinMax() {
+  }
+
+  {
+    // Max(5, Max(x, Max(y, Max(z, 8)))) => Max(Max(Max(x, 8), y), z)
```
I moved some of these into new macros and got rid of some code. But as we discussed offline, I am going to keep the rest as verbose as is now, so that test failures point to the correct test case.

navahgar commented on torch/csrc/jit/tensorexpr/ir_simplifier.cpp:
```diff
@@ -980,17 +994,138 @@ const Expr* PolynomialTransformer::mutate(const Max* v) {
+    return rhs_new;
+  }
+
+  auto combine_scalars = [&](const Expr* c1, const Expr* c2) -> const Expr* {
```
I tried doing that, but that introduces complications while finally generating code from `MinTerm` / `MaxTerm`. This is because, if the `scalar` is set to the minimum value at that point, that has to be ignored. I think keep it null is simpler.

Let me know if you strongly believe this should not be done.

navahgar commented on torch/csrc/jit/tensorexpr/ir_simplifier.cpp:
```diff
@@ -980,17 +994,138 @@ const Expr* PolynomialTransformer::mutate(const Max* v) {
+    }
+    if (scalar) {
+      return new MaxTerm(
+          hasher_, scalar, v->propagate_nans(), std::move(variables));
```
There was another bug due to which this was necessary earlier. Removed it now.

navahgar commented on torch/csrc/jit/tensorexpr/ir_simplifier.cpp:
```diff
@@ -1002,17 +1137,138 @@ const Expr* PolynomialTransformer::mutate(const Min* v) {
+    return rhs_new;
+  }
+
+  auto combine_scalars = [&](const Expr* c1, const Expr* c2) -> const Expr* {
```
Thanks for pointing this out. Refactored all the common parts into reusable template functions.

navahgar commented on torch/csrc/jit/tensorexpr/ir_simplifier.cpp:
```diff
@@ -980,17 +994,138 @@ const Expr* PolynomialTransformer::mutate(const Max* v) {
+    return combine_maxterms(lhs_maxterm, rhs_maxterm);
+  }
+
+  const Max* lhs_max = dynamic_cast<const Max*>(lhs_new);
```
While not when `propagate_nans` do not match, but in all other cases I am now converting `Max` to `MaxTerm` and `Min` to `MinTerm`, even when there are only 2 operands. That does simplify the code quite a bit. Thanks for the suggestion.

navahgar commented:
> Ok so functionally this looks great, and I like the test coverage (although I think you can commonalize a lot of the code), but I think we should refactor this before landing - there's too much copied code that can be eliminated. It's worth aiming for as clear and simple an implementation as possible here because the simplifier is a black box to most users and we want to smooth out any maintenance as much as we can.

Thanks for the review and useful suggestions. Addressed your comments. PTAL.

nickgg commented on torch/csrc/jit/tensorexpr/ir_simplifier.cpp:
```diff
@@ -1558,6 +1732,60 @@ const Expr* TermExpander::mutate(const Polynomial* v) {
+  }
+  for (size_t i = 1; i < variables.size(); i++) {
+    auto currNodeHash = v->hasher().hash(variables[i]);
+    if (currNodeHash != lastNodeHash) {
```
Feels much better to unique these when you add them to the OpTerm, not when you expand them.

navahgar replied:
Sure, good point. Done.

nickgg reviewed (commented):
Looks good generally, just one more change.

nickgg reviewed (approved):
LGTM! 

🚢
